### ISSUE TITLE
libopencv_dnn_modern.a build issue on Fedora 25 and CUDA 9

### ISSUE THREAD

**Issue by pstieber** [opencv/opencv#9789]
##### System information (version)

- OpenCV => master d25ee8a and opencv_contrib e80393b
- Operating System / Platform => Fedora 25 (4.12.14-200.fc25.x86_64) with CUDA 9.0
- Compiler => (GCC 6.4.1 20170727 (Red Hat 6.4.1-1)

##### Detailed description

I'm building the master branch of OpenCV with CUDA 9.0 and receiving many library creation errors due to multiply defined symbols:

```
[ 92%] Linking CXX shared library ../../lib/libopencv_java330.so
../../lib/libopencv_dnn_modern.a(caffe_converter.cpp.o): In function `caffe::BlobShape::SetCachedSize(int) const':
caffe_converter.cpp:(.text._ZNK5caffe9BlobShape13SetCachedSizeEi+0x0): multiple definition of `caffe::BlobShape::SetCachedSize(int) const'
../../lib/libopencv_dnn.a(caffe.pb.cc.o):caffe.pb.cc:(.text._ZNK5caffe9BlobShape13SetCachedSizeEi+0x0): first defined here
../../lib/libopencv_dnn_modern.a(caffe_converter.cpp.o): In function `caffe::BlobShape::CopyFrom(google::protobuf::Message const&)':
caffe_converter.cpp:(.text._ZN5caffe9BlobShape8CopyFromERKN6google8protobuf7MessageE+0x0): multiple definition of `caffe::BlobShape::CopyFrom(google::protobuf::Message const&)'
...
```

##### Steps to reproduce

Build with cmake using:

```
cmake \
    -D CMAKE_BUILD_TYPE=release \
    -D CMAKE_CXX_FLAGS="-std=c++11" \
    -D BUILD_opencv_python=OFF \
    -D BUILD_DOCS=OFF \
    -D BUILD_EXAMPLES=OFF \
    -D BUILD_SHARED_LIBS=OFF \
    -D BUILD_opencv_matlab=OFF \
    -D CMAKE_INSTALL_PREFIX="/usr/local/opencv" \
    -D WITH_OPENGL=ON \
    -D WITH_CUDA=ON \
    -D WITH_CUBLAS=ON \
    -D CUDA_GENERATION=Maxwell \
    -D CUDA_NVCC_FLAGS="--std=c++11 --expt-relaxed-constexpr" \
    -D OPENCV_EXTRA_MODULES_PATH=../opencv_contrib/modules \
    -D BUILD_opencv_legacy=OFF \
    ../opencv
```

Should I post this on opencv_contrib?

**Comment by pstieber** [comment]
I can use `-D BUILD_opencv_dnn_modern=OFF` to work around this, but still and wondering the cause of the multiply defined symbols.

**Comment by pstieber** [comment]
Also experiencing this using
- OpenCV => 3.3.0 tag
- Operating System / Platform => CentOS 7.4 (3.10.0-693.2.2.el7.x86_64) with CUDA 8.0
- Compiler => GCC 4.8.5 20150623 (Red Hat 4.8.5-16)

I do not experience the problem if I remove `-D BUILD_SHARED_LIBS=OFF`.  In other words, building shared libraries prevents the problem.  I want a static build of OpenCV so I add `-D BUILD_opencv_dnn_modern=OFF` to work around this, just like I did on Fedora.

**Comment by alalek** [comment]
Reproduced on master with these CMake flags:
```
-DBUILD_PROTOBUF=OFF -DPROTOBUF_UPDATE_FILES=ON -D BUILD_SHARED_LIBS=OFF
```

Problem is related to opencv_contrib module: dnn_modern

**Comment by bmstoyon** [comment]
It looks like the `caffe.proto` file exists in two places (`opencv/dnn` and `opencv_contrib/dnn_modern` via `tinydnn`).  Thus there are multiply generated `caffe.pb.cc` files and objects.

Here are links to the files:

`modules/dnn/src/caffe/caffe.proto`
https://github.com/opencv/opencv/blob/master/modules/dnn/src/caffe/caffe.proto

and

`tiny_dnn/io/caffe/caffe.proto`
https://github.com/tiny-dnn/tiny-dnn/blob/master/tiny_dnn/io/caffe/caffe.proto

**Comment by alalek** [comment]
caffe.proto from `opencv/dnn` is renamed (#10092). Please try code from the latest master.

**Comment by pstieber** [comment]
The build is fixed in the master branch at ce20efb8e7ef75f83a19c1b9b3976e6511cc002b.  Thanks!
